### PR TITLE
Optimizations for LinkedHashMap / LinkedHashSet

### PR DESCRIPTION
**Pull request by liang3zy22**
As Irytz suggested, I made this PR to continue optimize collection codes which are related with LinkedHashMap and LinkedHashSet.
There will be binary compatibility issue for the second commit. We can discuss it after all tests finished.

### CODE REVIEW

**Comment by liang3zy22** [comment]
As expected, test failed at binary compatibility. The error log is as below:
```
[error] 1 of 11 test tasks failed:
[error] - library/mimaReportBinaryIssues
[error]   - library/mimaReportBinaryIssues failed: java.lang.RuntimeException: Failed binary compatibility check against org.scala-lang:scala-library:2.13.10! Found 3 potential problems (filtered 8)
[error] Saved log of library/mimaReportBinaryIssues:
[error] [error] scala-library: Failed binary compatibility check against org.scala-lang:scala-library:2.13.10! Found 3 potential problems (filtered 8)
[error] [error]  * method addAll(scala.collection.IterableOnce)scala.collection.mutable.LinkedHashMap in class scala.collection.mutable.LinkedHashMap does not have a correspondent in other version
[error] [error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.mutable.LinkedHashMap.addAll")
[error] [error]  * method subtractAll(scala.collection.IterableOnce)scala.collection.mutable.LinkedHashMap in class scala.collection.mutable.LinkedHashMap does not have a correspondent in other version
[error] [error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.mutable.LinkedHashMap.subtractAll")
[error] [error]  * method filterInPlace(scala.Function2)scala.collection.mutable.LinkedHashMap in class scala.collection.mutable.LinkedHashMap does not have a correspondent in other version
[error] [error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.mutable.LinkedHashMap.filterInPlace")
[error] Failure due to previous errors
[error] (testAll1) Failure due to previous errors
[error] Total time: 455 s (07:35), completed Dec 5, 2022 3:47:15 AM

```
The method addAll,  subtractAll and filterInPlace reported binary compatibility issue. But the method getOrElse, getOrElseUpdate and hashCode() are also overridden and no binary compatibility issue. 
@lrytz, do you know why some methods have binary compatibility issue and others not? Is there any way to fix the binary compatibility issue?

**Comment by som-snytt** [comment]
@liang3zy22 methods from mixin traits are mixed-in, not inherited, so defining them would be binary compatible.

**Comment by liang3zy22** [comment]
@som-snytt , thanks for reply! The method getOrElseUpdate and filterInPlace are in the same trait, MapOps. But one is binary compatible and another not. don't know why the two methods are not same.

**Comment by som-snytt** [comment]
Hmm, these methods return `this.type`. That narrows the erased result in the signature.

Also, the concrete class extends `AbstractThing`. So actually I don't know how MiMa works.

I guess I don't understand erasure. You get a `Growable addAll(IterableOnce)` and a generic `LinkedHashMap<K, V> addAll(...)`. In `AbstractMap`, it is `Growable<Tuple2<K, V>> addAll(...)`.

**Comment by liang3zy22** [comment]
I also think so. The return type should cause the binary compatibility issue.

File: src/library/scala/collection/mutable/LinkedHashSet.scala [hunk 1]
```diff
@@ -270,6 +284,36 @@ class LinkedHashSet[A]
     }
   }
 
+    override def hashCode(): Int = {
```
Comment (lrytz) [inline src/library/scala/collection/mutable/LinkedHashSet.scala]: This seems to be a copy-past mistake, the set hashcode is computed differently. See `collection.Set.hashCode` and `mutable.HashSet.hashCode`.

File: src/library/scala/collection/mutable/LinkedHashSet.scala [hunk 2]
```diff
@@ -106,6 +107,17 @@ class LinkedHashSet[A]
       if (hasNext) { val res = cur.key; cur = cur.later; res }
       else Iterator.empty.next()
   }
+  private[collection] def entrysIterator: Iterator[Entry] = new AbstractIterator[Entry] {
```
Comment (lrytz) [inline src/library/scala/collection/mutable/LinkedHashSet.scala]: I think `entryIterator` sounds better (also in `LinkedHashMap`)

File: src/library/scala/collection/mutable/LinkedHashMap.scala [hunk 3]
```diff
@@ -110,6 +111,111 @@ class LinkedHashMap[K, V]
     case nd => Some(nd.value)
   }
 
+  override def getOrElse[V1 >: V](key: K, default: => V1): V1 = {
```
Comment (lrytz) [inline src/library/scala/collection/mutable/LinkedHashMap.scala:118]: Maybe we should keep the type test like in `mutable.HashMap` to not break existing subclasses

```
    if (getClass != classOf[LinkedHashMap[_, _]]) {
      // subclasses of HashMap might customise `get` ...
      super.getOrElse(key, default)
    } else {
      // .. but in the common case, we can avoid the Option boxing.
```

File: src/library/scala/collection/mutable/LinkedHashMap.scala [hunk 4]
```diff
@@ -110,6 +111,111 @@ class LinkedHashMap[K, V]
...
+      // .. but in the common case, we can avoid the Option boxing.
+      val nd = findEntry(key)
+      if (nd eq null) default else nd.value
+
+  }
+
+  override def getOrElseUpdate(key: K, defaultValue: => V): V = {
```
Comment (lrytz) [inline src/library/scala/collection/mutable/LinkedHashMap.scala:129]: also keep the type test?

**Comment by lrytz** [comment]
Adding an override with a refined return type is not forwards binary compatbile.

```
class A { def f: Object = "" }
class B extends A { override def f: String = "" }
```

In bytecode, the class `B` gets a method `public f()Ljava/lang/String;` and additionally `public synthetic bridge f()Ljava/lang/Object;`. In bytecode, overrides need to have exactly matching signatures, so the bridge is the actual override.

This happens when adding `override def addAll(xs: IterableOnce[(K, V)]): this.type` in `LinkedHashMap`: `this.type` erases to return type `LinkedHashMap` in bytecode, and a bridge is generated for overriding. The method with the new signature is not forwards binary compatible.

I tried to make this work a while ago but came across too many issues: https://github.com/scala/scala/pull/9141.

**Review by lrytz — CHANGES_REQUESTED** [review]

File: src/library/scala/collection/mutable/LinkedHashSet.scala (hunk 1 above)
Comment (liang3zy22) [reply src/library/scala/collection/mutable/LinkedHashSet.scala]: Done

File: src/library/scala/collection/mutable/LinkedHashSet.scala (hunk 2 above)
Comment (liang3zy22) [reply src/library/scala/collection/mutable/LinkedHashSet.scala]: Done

File: src/library/scala/collection/mutable/LinkedHashMap.scala (hunk 3 above)
Comment (liang3zy22) [reply src/library/scala/collection/mutable/LinkedHashMap.scala:118]: Done.

File: src/library/scala/collection/mutable/LinkedHashMap.scala (hunk 4 above)
Comment (liang3zy22) [reply src/library/scala/collection/mutable/LinkedHashMap.scala:129]: Done. I also added type test in updateWith like in mutable.HashMap.

**Comment by liang3zy22** [comment]
It seems that a lot common methods can't be overridden. I have tried concat, it also isn't binary compatible.

**Review by lrytz — APPROVED** [review]
Thanks for the improvements @liang3zy22. For the record, I submitted a proposal that would enable adding more performance overrides in the future: https://github.com/scala/improvement-proposals/pull/54

**Comment by liang3zy22** [comment]
Great. More optimizations can be realized for standard library if the SIP is accepted.

**Pull request merged**